### PR TITLE
Terminate active sessions at privileges removal of the collaborator user

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Add new Event Publisher Proxy creating event and publishing events related to role management `RoleManagementEventPublisherProxy`

### CODE REVIEW

File: components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+import org.wso2.carbon.identity.role.mgt.core.internal.RoleManagementServiceComponentHolder;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class handles creating event and publishing events related to role management.
+ */
+public class RoleManagementEventPublisherProxy {
```
Comment (emswbandara) [inline components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java:36]: add a new line after class declaration

File: components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    }
+
+    public static RoleManagementEventPublisherProxy getInstance() {
+
+        return proxy;
+    }
+
+    public void publishPreAddRole(String roleName, List<String> userList, List<String> groupList,
+                                  List<String> permissions, String tenantDomain) {
```
Comment (emswbandara) [inline components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java:50]: add new line after method signature.

Comment (emswbandara) [reply components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java:50]: please check other methods as well.

File: features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json [hunk 3]
```diff
@@ -170,7 +170,8 @@
   "identity_mgt.events.schemes.identityOauthEventHandler.module_index": "28",
   "identity_mgt.events.schemes.identityOauthEventHandler.subscriptions": [
     "POST_SET_USER_CLAIMS",
-    "POST_SET_USER_CLAIM"
+    "POST_SET_USER_CLAIM",
+    "POST_UPDATE_USER_LIST_OF_ROLE"
```
Comment (emswbandara) [inline features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json:174]: we will have to revoke tokens/sessions on delete role, permission update as well.

Comment (Thisara-Welmilla) [reply features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json]: Addressed with commit f0f156e8e8990df3af44922e65acf1061762fda5

File: components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java:36]: Addressed with commit f0f156e8e8990df3af44922e65acf1061762fda5

File: components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementEventPublisherProxy.java:51]: Addressed with commit f0f156e8e8990df3af44922e65acf1061762fda5

File: components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java [hunk 4]
```diff
@@ -72,8 +72,24 @@ private Event(){}
...
+        public static final String PRE_UPDATE_ROLE_NAME = "PRE_UPDATE_ROLE_NAME";
+        public static final String POST_UPDATE_ROLE_NAME = "POST_UPDATE_ROLE_NAME";
         public static final String PRE_DELETE_ROLE= "PRE_DELETE_ROLE";
         public static final String POST_DELETE_ROLE= "POST_DELETE_ROLE";
+        public static final String PRE_GET_USER_LIST_OF_ROLE = "PRE_GET_USER_LIST_OF_ROLE";
+        public static final String POST_GET_USER_LIST_OF_ROLE = "POST_GET_USER_LIST_OF_ROLE";
+        public static final String PRE_GET_GROUP_LIST_OF_ROLES = "PRE_GET_GROUP_LIST_OF_ROLES";
+        public static final String POST_GET_GROUP_LIST_OF_ROLES = "POST_GET_GROUP_LIST_OF_ROLES";
+        public static final String PRE_UPDATE_GROUP_LIST_OF_ROLE = "PRE_UPDATE_GROUP_LIST_OF_ROLE";
+        public static final String POST_UPDATE_GROUP_LIST_OF_ROLE = "UPDATE_GROUP_LIST_OF_ROLE";
```
Comment (emswbandara) [inline components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java:88]: shouldn't the value of this be POST_UPDATE_GROUP_LIST_OF_ROLE ?

File: features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json [hunk 5]
```diff
@@ -170,7 +170,10 @@
   "identity_mgt.events.schemes.identityOauthEventHandler.module_index": "28",
   "identity_mgt.events.schemes.identityOauthEventHandler.subscriptions": [
     "POST_SET_USER_CLAIMS",
-    "POST_SET_USER_CLAIM"
+    "POST_SET_USER_CLAIM",
+    "POST_UPDATE_USER_LIST_OF_ROLE",
+    "PRE_DELETE_ROLE",
```
Comment (emswbandara) [inline features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json:175]: POST_DELETE_ROLE?

File: components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java]: Addressed with commit 24f0165df0577d035bb0473ae2094d5a70751997

File: features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json (hunk 5 above)
Comment (Thisara-Welmilla) [reply features/identity-event/org.wso2.carbon.identity.event.server.feature/resources/org.wso2.carbon.identity.event.server.feature.default.json:175]: Have to go with the `PRE_DELETE_ROLE` as after role deletion, cannot retrieve the user Id list to revoke tokens and terminate sessions.

**Review by emswbandara — APPROVED** [review]

**Comment by Thisara-Welmilla** [comment]
Reopened with https://github.com/wso2/carbon-identity-framework/pull/3675

**Pull request closed without merging**